rfcs/rdm-0064-custom-fields.md

```diff
@@ -0,0 +1,440 @@
+}
+```
+
+#### Custom field type / Marshmallow / Elasticsearch chart
+
+The following table represents the type conversion/relation between custom field type, Marshmallow fields and Elasticsearch mapping.
+
+| Custom Field type | Marshmallow field | Elasticsearch mapping |
+| :---------------: | :---------------: | :-------------------: |
+|      string       | SanitizedUnicode  |         text          |
+|      keyword      | SanitizedUnicode  |        keyword        |
+|       date        |     DateField     |    data isoformat     |
+|      integer      |      Integer      |        integer        |
+|      double       |      Number       |        double         |
+|       list        |       List        |   inner value type    |
+|       dict        |       Dict        |        object         |
+
+#### How is it linked to a record
+
+> Note: in this sub section the term _record_ is treated as a generic term, namely a document. It will refer to bibliographic and community records as _records_.
+
+Custom vocabularies are added to the record's data model and dumped in their json blob. However, vocabularies are different since they need to be linked. Meaning, they need to be checked for existence when a record is created and dereferenced when it is read.
+
+The `Record` API stays compatible with its previous behaviour, accessing relation fields through the `record.relations.myrelationfield`. However, the previous `RelationsField` has been changed to `MultiRelationsField`, which accepts implementations of `RelationBase` or `RelationsField`. This allows all custom fields to be injected in the relations attribute.
+
+While the code looks similar to the following snippet, when accessing the attributes they would not need access the `custom` relation. The custom fields relations are injected by the `CustomFieldsRelation`. For example, accessing the previously configured `experiments` custom field would be done as `record.relations.experiments`, while `record.relations.custom` would give an attribute error.
+
+```python
+class Record:
+    relations = MultiRelationsField(
+        custom=CustomFieldsRelation(),
+    )
+```
+
+The dereferencing part is transparent, i.e. no code needs to be altered since the API is still the same. It is carried out by the `RelationsComponent` which will dereference the relations.
+
+### Creating a _custom field_
+
+Once the custom fields are configured in the `invenio.cfg` file, they need to be added to the record's mapping, so their values are properly indexed in Elasticsearch. That can be done using the CLI.
+
+**create**
+
+You can add fields to the records and community data models using the following command. You can pass any number of configured fields.
+
+```shell
+invenio custom-fields create [-f <fieldname>] [-f <fieldname>]
+```
+
+In order to create _all_ configured values, just ignore the `-f` option. For example, on the first time an instance is set up.
+
+```shell
+invenio custom-fields create
+```
+
```

Comment:
do you want to add anything that, if the type changes, an error is thrown?

Reply:
Hmm I avoided it since it was explained above in the limitation. Do you find it necessary? I can add it if so.